### ISSUE TITLE
How to deal with the conv in SE module？

### ISSUE THREAD

**Issue by RainFrost1** [keyu-tian/SparK#54]

**Comment by will1973** [comment]
same question

**Comment by yxchng** [comment]
any updates?

**Comment by RainFrost1** [comment]
Looking forward your reply, any suggestions? @keyu-tian

**Comment by keyu-tian** [comment]
@RainFrost1 @will1973 @yxchng SE module contains two linear layers (channel-wise) and one global average pooling. No changes required to the linear layers. As for the global pooling on a masked feature map, one should calculate the mean value of non-masked positions only. To implement this you can use our function `_get_active_ex_or_ii` in https://github.com/keyu-tian/SparK/blob/main/pretrain/encoder.py#L30 like:

```python
class SparseGlobalAveragePooling(nn.Module):
    def forward(self, x):   # shape: BCHW
        B, C, H, W = x.shape
        non_masked_positions = _get_active_ex_or_ii(H=H, W=W, returning_active_ex=True)
        mean = (x * non_masked_positions).sum(dim=(2,3), keepdims=True) / non_masked_positions.sum(dim=(2,3), keepdims=True)
        return mean         # shape: BC11
```
